doc/designs/ech-api.md

```diff
@@ -0,0 +1,593 @@
+of version code-points which have been tracking Internet-Draft version
+numbers during the course of spec development.
+
+The only current ECHConfig version supported is 0xfe0d which will be the
+value to be used in the eventual RFC when that issues. (We'll replace the
+XXXX with the relevant RFC number once that's known.)
+
+```c
+/* version from RFC XXXX */
+#  define OSSL_ECH_RFCXXXX_VERSION 0xfe0d
+/* latest version from an RFC */
+#  define OSSL_ECH_CURRENT_VERSION OSSL_ECH_RFCXXXX_VERSION
+```
+
+Note that 0xfe0d is also the value of the ECH extension codepoint:
+
+```c
+#  define TLSEXT_TYPE_ech                       0xfe0d
+```
+
+The uses of those should be correctly differentiated in the implementation, to
+more easily avoid problems if/when new versions are defined.
+
+"GREASEing" is defined in
+[RFC8701](https://datatracker.ietf.org/doc/html/rfc8701) and is a mechanism
+intended to discourage protocol ossification that can be used for ECH.  GREASEd
+ECH may turn out to be important as an initial step towards widespread
+deployment of ECH.
+
+Minimal Sample Code
+-------------------
+
+OpenSSL includes code for an
+[``sslecho``](https://github.com/sftcd/openssl/tree/ECH-draft-13c/demos/sslecho)
+demo.  We've added a minimal
+[``echecho``](https://github.com/sftcd/openssl/blob/ECH-draft-13c/demos/sslecho/echecho.c)
+that shows that adding one new server call
+(``SSL_CTX_ech_enable_server_buffer()``) and one new client call
+(``SSL_CTX_ech_set1_echconfig()``) is all that's needed to ECH-enable this
+demo.
+
+Handling Custom Extensions
+--------------------------
+
+OpenSSL supports custom extensions (via ``SSL_CTX_add_custom_ext()``) so that
+extension values are supplied and parsed by client and server applications via
+a callback.  The ECH specification of course doesn't deal with such
+implementation matters, but comprehensive ECH support for such custom
+extensions could quickly become complex. At present, in the absence of evidence
+of sensitive custom extension values, we handle all such extensions by using
+the ECH compression mechanism.  That means we require no API changes, only make
+one call to the application callbacks and get interoperability, but that such
+extension values remain visible to network observers. That could change if some
+custom value turns out to be sensitive such that we'd prefer to not include it
+in the outer CH.
+
+Server-side APIs
+----------------
+
+The main server-side APIs involve generating a key and the related
+ECHConfigList structure that ends up published in the DNS, periodically loading
+such keys into a server to prepare for ECH decryption and handling so-called
+ECH split-mode where a server only does ECH decryption but passes along the
+inner CH to another server that does the actual TLS handshake with the client.
+
+### Key and ECHConfigList Generation
+
+``ossl_edch_make_echconfig()`` is for use by command line or other key
```

Comment:
there's a typo (edch) yep, but should it be ``OSSL_ech_make_echconfig()`` or ``OSSL_ECH_make_echconfig()`` - I thought the former was right and the latter implied there was a ``typedef OSSL_ECH`` somewhere?

The backticks make the name render as ``code`` in the markdown, which seems nicer, but no objection to not doing that.

Reply:
`OSSL_ECH_` prefix does not necessarily imply that there is an `OSSL_ECH` type. ECH can also be a subsystem.

As for the backticks - sure, there should be backticks but I wanted to know why do you use double backticks instead of single.

Reply:
So I take it the right thing is to change from `OSSL_ech_make_echconfig()` to `OSSL_ECH_make_echconfig()` - will do that now.

I dunno why I use double backticks, maybe some other flavour of markdown needed it and it's muscle-memory, not sure tbh:-) If one is enough, that's fine though.